rsqueakvm/storage_contexts.py

```diff
@@ -472,35 +495,70 @@ def stacksize(self):
 
     def full_stacksize(self):
         if not self.pure_is_block_context():
-            # Magic numbers... Takes care of cases where reflective
-            # code writes more than actual stack size
             method = self.w_method()
             assert isinstance(method, W_CompiledMethod), "context method is not a W_CompiledMethod"
-            stacksize = method.compute_frame_size()
+            stacksize = method.frame_size()
         else:
-            # TODO why not use method.compute_frame_size for BlockContext too?
             stacksize = self.stacksize()  # no temps
         return stacksize
 
     def init_temps_and_stack(self):
         self = fresh_virtualizable(self)
         stacksize = self.full_stacksize()
-        self._temps_and_stack = [None] * stacksize
+        self._temps_and_stack = [self.space.w_nil] * stacksize
         tempsize = self.tempsize()
         self.store_stack_ptr(tempsize)  # we point after the last element
 
     def stack_get(self, index0):
         assert index0 >= 0, "trying to stack_get negative index"
-        return self._temps_and_stack[index0] or self.space.w_nil
+        if self.update_stacksize(index0):
+            return self.overflow_stack_get(index0)
+        return self._temps_and_stack[index0]
+
+    def overflow_stack_get(self, index0):
+        overflow_index = index0 - len(self._temps_and_stack)
+        return self.get_extra_data()._overflow_stack[overflow_index] or self.space.w_nil
 
     def stack_put(self, index0, w_val):
         assert w_val is not None, "trying to put None on the stack"
         assert index0 >= 0, "trying to stack_put at negative index"
+        if self.update_stacksize(index0):
+            return self.overflow_stack_put(index0, w_val)
         self._temps_and_stack[index0] = w_val
 
+    def overflow_stack_put(self, index0, w_val):
+        overflow_index = index0 - len(self._temps_and_stack)
+        self.get_extra_data()._overflow_stack[overflow_index] = w_val
+
+    def has_overflow_stack(self):
+        extra_data = self.extra_data()
+        if extra_data:
+            return extra_data._overflow_stack is not None
+        return False
+
+    @jit.unroll_safe
+    def update_stacksize(self, index0):
```

Comment:
We cache on the compiled method how large the largest frame was that we needed for this method